homeworks/solutions/krasnor/2-ddg/src/halfedge.js

```diff
@@ -0,0 +1,634 @@
+        for (let h of outHfEdges.values()) {
+            // area += h.face.getArea()/3; // area approximation
+
+            // voronoi vertex area
+            const u = h.prev.getVector().norm();
+            const v = h.getVector().norm();
+            area += (u * u * h.prev.cotan() + v * v * h.cotan()) / 8
+        }
+
+        let c_sum = new Vector();
+        for (let hfe of outHfEdges.values()) {
+            //     O--------i
+            //      \ A   /  \
+            //       \   hfe  \
+            //        \ /     B\
+            //         J--------O
+
+            let fac = hfe.cotan() + hfe.twin.cotan();
+
+            c_sum = c_sum.sub(  hfe.getVector().scale(fac)  );
+        }
+
+        return c_sum.scale( 1/(2*area)); // return normal
+    }
+
+    /**
+     * Calculates the Discrete Gaussian Curvature Value
+     *
+     * @returns {number} Curvature value between 0 and 1 (360°-0°)
+     */
+    calcGaussianCurvature() {
+        let outgoingHfEdges = this.getAllOutgoingHalfedges();
+        let gauss_curv_sum = 0.0;
+
+        for (let hfe of outgoingHfEdges.values()) {
+            let hf0 = hfe;
+            let hf2 = hfe.prev.twin;
+
+            let angle_dot = (hf0.getVector().unit().dot(hf2.getVector().unit()));
+            let theta_i = Math.acos(angle_dot);
+            // let theta_deg = theta_i * (180 / Math.PI);
+
+            gauss_curv_sum += theta_i;
+        }
+
+        return (2 * Math.PI) - gauss_curv_sum;
+    }
+
+    /**
+     * numAdjacentEdges returns the number of adjacency edges of the given vertex
+     * @param vertex is an vertex from a halfedge-based mesh
+     * @returns {Number}
+     */
+    numAdjacentEdges() {
+        const e0 = this.halfedge
+        let edge_indices = [e0.idx]
+        for (let e = e0.twin.next; e != e0; e = e.twin.next) { // FIXME works not for boundraries
+            if (e == null || e == undefined) // boundary reached
+                return edge_indices.length;
+            edge_indices.push(e.idx)
+        }
+        return edge_indices.length
+    }
+
+    /**
+     * Returns all adjacent Faces to this vertex
+     * @returns {IterableIterator<any>}
+     */
+    getAdjacentFaces() {
+        const e0 = this.halfedge;
+        let faces = new Map();
+        faces.set(this.halfedge.face.idx, this.halfedge.face);
+        for (let e = e0.twin.next; e != e0; e = e.twin.next) {
+            if (e == null || e == undefined || e.face == null || e.face == undefined) {
+                break;
+            }
+            if (!faces.has(e.face.idx))
+                faces.set(e.face.idx, e.face);
+        }
+        return faces.values();
+    }
+
+    /**
+     * Returns a Map of all Halfedges that point away from this vertex. Key is the corresponding face index
+     *
+     * @returns {Map<Number, Halfedge>}
+     */
+    getAllOutgoingHalfedges() {
+        const e0 = this.halfedge;
+        let hfedges = new Map();
```

Comment:
Nit: This could cause a lot of allocation when you loop through all vertices. Check the provided reference solution might give you a few inspirations about how to optimize here.